aggregator.go

```diff
@@ -0,0 +1,300 @@
+func Aggregate(ds ipldformat.DAGService, toAggregate []AggregateDagEntry) (aggregateRoot cid.Cid, err error) {
+
+	dags := make(map[string]*ManifestDagEntry, len(toAggregate))
+
+	for _, d := range toAggregate {
+
+		cv1 := cid.Undef
+		if d.RootCid.Version() == 1 {
+			cv1 = d.RootCid
+		} else {
+			cv1 = cid.NewCidV1(d.RootCid.Type(), d.RootCid.Hash())
+		}
+
+		// do not consider identity multihashes for aggregation
+		if cv1.Prefix().MhType == uint64(multicodec.Identity) {
+			continue
+		}
+
+		cv0 := cid.Undef
+		if cv1.Prefix().Codec == uint64(multicodec.DagPb) &&
+			cv1.Prefix().MhType == uint64(multicodec.Sha2_256) &&
+			cv1.Prefix().MhLength == 32 {
+			cv0 = cid.NewCidV0(cv1.Hash())
+		}
+
+		dagName := cv1.String()
+
+		if _, exists := dags[dagName]; !exists {
+			dags[dagName] = &ManifestDagEntry{
+				rootCid:    cv1,
+				RecordType: DagAggregateEntry,
+				DagCidV1:   dagName,
+			}
+			if cv0 != cid.Undef {
+				dags[dagName].DagCidV0 = cv0.String()
+			}
+			if d.UniqueBlockCumulativeSize != 0 {
+				v := d.UniqueBlockCumulativeSize
+				dags[dagName].DagSize = &v
+			}
+			if d.UniqueBlockCount != 0 {
+				v := d.UniqueBlockCount
+				dags[dagName].NodeCount = &v
+			}
+		}
+	}
+
+	if len(dags) == 0 {
+		return cid.Undef, xerrors.New("no valid entries to aggregate")
+	}
+
+	// unixfs sorts internally, so we need to pre-sort to match up insertion indices
+	sortedDags := make([]*ManifestDagEntry, 0, len(dags))
+	for _, d := range dags {
+		sortedDags = append(sortedDags, d)
+	}
+	sort.Slice(sortedDags, func(i, j int) bool {
+		return strings.Compare(sortedDags[i].DagCidV1, sortedDags[j].DagCidV1) < 0
+	})
+
+	// innermost layer, 4 bytes off end
+	innerNodes := make(nodeMap)
+	for _, d := range sortedDags {
+
+		parentName := d.DagCidV1[:3] + `...` + d.DagCidV1[len(d.DagCidV1)-4:]
+		if _, exists := innerNodes[parentName]; !exists {
+			innerNodes[parentName] = emptyDir()
+		}
+
+		var dagSize uint64
+		if d.DagSize != nil {
+			dagSize = *d.DagSize
+		}
+
+		if err := innerNodes[parentName].AddRawLink(d.DagCidV1, &ipldformat.Link{
+			Size: dagSize,
+			Cid:  d.rootCid,
+		}); err != nil {
+			return cid.Undef, err
+		}
+		d.PathIndexes[2] = len(innerNodes[parentName].Links()) - 1
+	}
+
+	newBlocks := make([]ipldformat.Node, 0, 1<<20)
+
+	// secondary layer, 2 bytes off end ( drop the 2 second-to-last )
+	outerNodes := make(nodeMap)
+	for _, nodeName := range sortedNodeNames(innerNodes) {
+		nd := innerNodes[nodeName]
+
+		newBlocks = append(newBlocks, nd)
+		parentName := nodeName[0:len(nodeName)-4] + nodeName[len(nodeName)-2:]
+		if _, exists := outerNodes[parentName]; !exists {
+			outerNodes[parentName] = emptyDir()
+		}
+
+		if err := outerNodes[parentName].AddNodeLink(nodeName, nd); err != nil {
+			return cid.Undef, err
+		}
+
+		for _, innerDaglink := range nd.Links() {
+			d := dags[innerDaglink.Name]
+			d.PathPrefixes[1] = nodeName
+			d.PathIndexes[1] = len(outerNodes[parentName].Links()) - 1
```

Comment:
Nit: same possible usage of ignored index `_, exists`?

Reply:
Same as https://github.com/filecoin-project/go-dagaggregator-unixfs/pull/1#discussion_r660977261